sw/device/tests/crypto/rsa_4096_signature_functest.c

```diff
@@ -0,0 +1,323 @@
+                                    const uint32_t *sig,
+                                    const rsa_padding_t padding_mode,
+                                    const rsa_hash_t hash_mode,
+                                    hardened_bool_t *verification_result) {
+  key_mode_t key_mode;
+  switch (padding_mode) {
+    case kRsaPaddingPkcs:
+      key_mode = kKeyModeRsaSignPkcs;
+      break;
+    case kRsaPaddingPss:
+      key_mode = kKeyModeRsaSignPss;
+      break;
+    default:
+      return INVALID_ARGUMENT();
+  };
+
+  rsa_public_key_t public_key = {
+      .n = {.key_mode = key_mode,
+            .key_length = sizeof(kTestModulus),
+            .key = kTestModulus,
+            .checksum = 0},
+      .e = {.key_mode = key_mode,
+            .key_length = sizeof(kTestPublicExponent),
+            .key = &kTestPublicExponent,
+            .checksum = 0},
+  };
+
+  public_key.n.checksum = integrity_unblinded_checksum(&public_key.n);
+  public_key.e.checksum = integrity_unblinded_checksum(&public_key.e);
+
+  crypto_const_byte_buf_t msg_buf = {
+      .data = msg,
+      .len = msg_len,
+  };
+
+  crypto_const_word32_buf_t sig_buf = {
+      .data = sig,
+      .len = kRsa4096NumWords,
+  };
+
+  uint64_t t_start = profile_start();
+  TRY(otcrypto_rsa_verify(&public_key, msg_buf, padding_mode, hash_mode,
+                          sig_buf, verification_result));
+  profile_end_and_print(t_start, "RSA verify");
+
+  return OK_STATUS();
+}
+
+status_t pkcs1v15_sign_test(void) {
+  // Generate a signature using PKCS#1 v1.5 padding and SHA-512 as the hash
+  // function.
+  uint32_t sig[kRsa4096NumWords];
+  TRY(run_rsa_4096_sign(kTestMessage, kTestMessageLen, kRsaPaddingPkcs,
+                        kRsaHashSha512, sig));
+
+  // Compare to the expected signature.
+  TRY_CHECK_ARRAYS_EQ(sig, kValidSignaturePkcs1v15,
+                      ARRAYSIZE(kValidSignaturePkcs1v15));
+  return OK_STATUS();
+}
+
+status_t pkcs1v15_verify_valid_test(void) {
+  // Try to verify a valid signature.
+  hardened_bool_t verification_result;
+  TRY(run_rsa_4096_verify(kTestMessage, kTestMessageLen,
+                          kValidSignaturePkcs1v15, kRsaPaddingPkcs,
+                          kRsaHashSha512, &verification_result));
+
+  // Expect the signature to pass verification.
+  TRY_CHECK(verification_result == kHardenedBoolTrue);
+  return OK_STATUS();
+}
+
+status_t pkcs1v15_verify_invalid_test(void) {
+  // Try to verify an invalid signature (wrong padding mode).
+  hardened_bool_t verification_result;
+  TRY(run_rsa_4096_verify(kTestMessage, kTestMessageLen, kValidSignaturePss,
+                          kRsaPaddingPkcs, kRsaHashSha512,
+                          &verification_result));
+
+  // Expect the signature to fail verification.
+  TRY_CHECK(verification_result == kHardenedBoolFalse);
+  return OK_STATUS();
+}
+
+OTTF_DEFINE_TEST_CONFIG();
+
+// Holds the test result.
+static volatile status_t test_result;
```

Comment:
does this have to be `static volatile`? can this just be a local variable in `test_main()`?

Reply:
No, it doesn't have to be, it's there for bad reasons. (I found this in some example and copied it, then later realized it could just be in `test_main()`, so some tests have it this way and some have it as a local variable. The RSA-2048 test had it this way and was the template for these tests.) Now fixed in all RSA tests.